src/library/scala/collection/mutable/LinkedHashMap.scala

```diff
@@ -129,6 +132,59 @@ class LinkedHashMap[K, V]
     }
   }
 
+  override def filterInPlace(p: (K, V) => Boolean): this.type = {
+    if (size > 0) {
+      var bucket = 0 // the index of the current bucket in the hash table
+      val tableTable = table.table
+      val buckets = tableTable.length
+      var newFirstEntry: Entry = null // will become the new firstEntry
+      var earlier: Entry = null // the entry "before" the current entry, may cross bucket boundaries
+      var newTableSize = table.tableSize
+
+      while (bucket < buckets) {
+        (tableTable(bucket) : @switch) match {
+          case null => ()
+          case head =>
+            var newHead: Entry = null // the new beginning of this bucket
+          var prev: Entry = null // the previous node within this bucket, does not cross bucket boundaries
+          var next: Entry = head.asInstanceOf[Entry] // the next node within this bucket, does not cross bucket boundaries
+
+            while (next ne null) {
+              if (p(next.key, next.value)) {
+                if (newHead eq null) {
+                  newHead = next
+                }
+                if (newFirstEntry eq null) {
+                  newFirstEntry = next
+                }
+                next.earlier = earlier // link up this node with previous node that passed filter
+                if (earlier ne null) earlier.later = next
+                earlier = next
+                prev = next
+              } else {
+                newTableSize -= 1
+              }
+              next = next.next
+            }
+
+            if (newHead ne head) tableTable(bucket) = newHead
+        }
+        bucket += 1
+      }
+      table.tableSize = newTableSize
+      firstEntry = newFirstEntry
+      lastEntry = earlier
+      if (earlier ne null) {
+        // Rather than set each node's `later` pointer to `next.next` as we iterate, we wait until we find the next
+        // node that passes the filter, then go back to `earlier` and set it's `later` to that node.
+        // However that means the last node that passes the filter will still point to its previous tail (if it wasn't
+        // the last element of the Map). Now at the end we make sure to cut that link.
+        earlier.later = null
```

Comment:
This doesn't sound right.  You're going through things in hash order, not insertion order, right?  So you have got to heal the pointers as you go.